crates/runc-shim/src/cgroup_memory.rs

```diff
@@ -0,0 +1,128 @@
+    {
+        if line.contains("memory") {
+            let parts: Vec<&str> = line.split(':').collect();
+            if let Some(path) = parts.last() {
+                return Ok(path.to_string());
+            }
+            break;
+        }
+    }
+    Err("Memory cgroup path not found").map_err(other_error!(e, "Error get path:"))?
+}
+
+pub async fn get_existing_cgroup_mem_path(pid_path: String) -> Result<(String, String)> {
+    let mut mountinfo_path = get_path_from_mountinfo().await?;
+    if mountinfo_path.0 == "/" {
+        mountinfo_path.0 = String::from("");
+    }
+    let res = pid_path.trim_start_matches(&mountinfo_path.0).to_string();
+    Ok((res, mountinfo_path.1))
+}
+
+async fn get_path_from_mountinfo() -> Result<(String, String)> {
+    let mountinfo_file = File::open("/proc/self/mountinfo".clone())
+        .await
+        .map_err(io_error!(e, "open {}", "/proc/self/mountinfo"))?;
+
+    // for line in BufReader::new(mountinfo_file).lines() {
+    let mut lines = BufReader::new(mountinfo_file).lines();
+    while let Some(line) = lines
+        .next_line()
+        .await
+        .map_err(io_error!(e, "read file /proc/self/mountinfo failed."))?
+    {
+        if line.contains("cgroup") && line.contains("memory") {
+            return parse_memory_mountroot(&line);
+        }
+    }
+
+    Err("No memory found in mountinfo file.")
+        .map_err(other_error!(e, "Error get path from montinfo:"))?
+}
+
+fn parse_memory_mountroot(line: &str) -> Result<(String, String)> {
+    let s_values: Vec<_> = line.split(" - ").collect();
+    if s_values.len() != 2 {
+        return Err("Parse from mountinfo failed")
+            .map_err(other_error!(e, "Error get path from montinfo:"))?;
+    }
+
+    let s0_values: Vec<_> = s_values[0].trim().split(' ').collect();
+    if s0_values.len() < 6 {
+        return Err("Parse from mountinfo failed")
+            .map_err(other_error!(e, "Error get path from montinfo:"))?;
+    }
+    Ok((s0_values[3].to_string(), s0_values[4].to_string()))
+}
+
+pub async fn register_memory_event(
+    key: &str,
+    cg_dir: &Path,
+    event_name: &str,
+) -> Result<Receiver<String>> {
+    let path = cg_dir.join(event_name);
+    let event_file = fs::File::open(path.clone())
+        .await
+        .map_err(other_error!(e, "Error get path:"))?;
+
+    // let eventfd = unsafe { libc::eventfd(0, libc::EFD_CLOEXEC) };
+    let eventfd = eventfd::eventfd(0, eventfd::EfdFlags::EFD_CLOEXEC)?;
+
+    let event_control_path = cg_dir.join("cgroup.event_control");
+    let data = format!("{} {}", eventfd.as_raw_fd(), event_file.as_raw_fd());
+    fs::write(&event_control_path, data.clone())
+        .await
+        .map_err(other_error!(e, "Error write eventfd:"))?;
+
+    let mut buf = [0u8; 8];
+
+    let (sender, receiver) = mpsc::channel(128);
+    let key = key.to_string();
+
+    tokio::spawn(async move {
+        let mut eventfd_file = unsafe { File::from_raw_fd(eventfd.as_raw_fd()) };
```

Comment:
Can you find another method to wrapper fd as a File? I think nix crate has a lot of safe function. Use unsafe function if there is only one choice.

Reply:
I find two save functions in nix lib, [Epoll](https://docs.rs/nix/0.27.1/src/nix/sys/epoll.rs.html#105-189) and [read](https://docs.rs/nix/0.27.1/src/nix/unistd.rs.html#1107-1113). 
For Epoll, in this scenario, only one eventfd needs to be monitored, it seems that read directly seems more lightweight.
For read func in lib nix, its implementation is alse unsfe:
```rust
pub fn read(fd: RawFd, buf: &mut [u8]) -> Result<usize> {
    let res = unsafe {
        libc::read(fd, buf.as_mut_ptr() as *mut c_void, buf.len() as size_t)
    };

    Errno::result(res).map(|r| r as usize)
}
```
In addition, these two funcs are both sync.

Reply:
@Burning1020 Hi, i don't find a suitable way in lib nix to wrap a fd as a file. I just find two ways to avoid using `unsafe`, comparation is discribed above, using `unsafe { File::from_raw_fd())` seems better. Take a look here?

Reply:
> @Burning1020 Hi, i don't find a suitable way in lib nix to wrap a fd as a file. I just find two ways to avoid using `unsafe`, comparation is discribed above, using `unsafe { File::from_raw_fd())` seems better. Take a look here?

No prob.